### PR TITLE
blockcms becomes LinkWidget

### PR DESCRIPTION
**Pull request by julienbourdeau**
This module provide simple link list for your theme. You can hook a widget to any display hook.

## Screenshots

link_widget_ _argarga_edegdth
link_widget_ _argarga_edegdth

## Smarty

In Smarty you get:

```
Array
(
    [0] => Array
        (
            [id] => 2
            [title] => www
            [hook] => displayFooter
            [position] => 1
            [links] => Array
                (
                    [0] => Array
                        (
                            [id] => cms-page-1
                            [class] => cms-page-link
                            [title] => Livraison
                            [description] => Nos conditions de livraison
                            [url] => http://prestashop.ps/fr/content/1-livraison
                        )

                    [1] => Array
                        (
                            [id] => cms-page-2
                            [class] => cms-page-link
                            [title] => Mentions légales
                            [description] => Mentions légales
                            [url] => http://prestashop.ps/fr/content/2-mentions-legales
                        )

                    [2] => Array
                        (
                            [id] => cms-page-3
                            [class] => cms-page-link
                            [title] => Conditions d'utilisation
                            [description] => Nos conditions d'utilisation
                            [url] => http://prestashop.ps/fr/content/3-conditions-utilisation
                        )

                    [3] => Array
                        (
                            [id] => cms-page-4
                            [class] => cms-page-link
                            [title] => A propos
                            [description] => En savoir plus sur notre entreprise
                            [url] => http://prestashop.ps/fr/content/4-a-propos
                        )

                    [4] => Array
                        (
                            [id] => cms-page-5
                            [class] => cms-page-link
                            [title] => Paiement sécurisé
                            [description] => Notre méthode de paiement sécurisé
                            [url] => http://prestashop.ps/fr/content/5-paiement-securise
                        )

                    [5] => Array
                        (
                            [id] => cms-page-6
                            [class] => cms-page-link
                            [title] => noouvelle page
                            [description] => cool desc
                            [url] => http://prestashop.ps/fr/content/6-efsef-se-g-seg-se-g
                        )

                )

        )

)
```

### TODO

- [x] Remove legacy code :trollface: 
- [x] Move from _configure page_ to _ModuleAdminController_ 
- [x] Implement WidgetInterface module
- [x] Create very simple template
- [x] Introduce Presenter
- [ ] Rename variable and classes to get rid of `CmsBlock` (use `LinkWidget` or `LinkGroup` instead
- [ ] Remove code from Repository (like `getCmsPages()`) ?
- [ ] Move `getDisplayHooksForHelper` to HookRepository
- [ ] Add many pages like there was in blockcms footer (stores, best price, price drop)
- [ ] **Allow user to reorder links**
- [ ] validate form
- [ ] Make sure everything is translatable

### CODE REVIEW

File: blockcms.php [hunk 1]
```diff
@@ -25,906 +25,146 @@
...
+        $this->author = 'PrestaShop';
+        $this->need_instance = 0;
+
+        $this->bootstrap = true;
+        parent::__construct();
+
+        $this->displayName = $this->l('CMS block');
+        $this->description = $this->l('Adds a block with several CMS links.');
+        $this->secure_key = Tools::encrypt($this->name);
+        $this->ps_versions_compliancy = array('min' => '1.6', 'max' => _PS_VERSION_);
```
Comment (maximebiloe) [inline blockcms.php]: 1.7 ?

File: blockcms.php [hunk 2]
```diff
@@ -25,906 +25,146 @@
...
+    protected $_html;
+    protected $_display;
+    private $cmsBlockPresenter;
+    private $cmsBlockRepository;
+
+    public function __construct()
+    {
+        $this->name = 'blockcms';
+        $this->tab = 'front_office_features';
+        $this->version = '2.1.1';
```
Comment (maximebiloe) [inline blockcms.php]: 3.0 ?

File: controllers/admin/AdminLinkWidgetController.php [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                'newBlock' => array(
+                    'title' => $this->module->l('New block'),
+                    'href' => $this->context->link->getAdminLink('AdminLinkWidget').'&amp;addCmsBlock',
+                    'class' => 'pull-right',
+                    'icon' => 'process-icon-new'
+                ),
+            ),
+        );
+
+        $this->getLanguages();
```
Comment (mickaelandrieu) [inline controllers/admin/AdminLinkWidgetController.php]: I dont understand what this function does ...

File: controllers/admin/AdminLinkWidgetController.php [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        $helper->allow_employee_form_lang = $this->allow_employee_form_lang;
+        $helper->toolbar_scroll = true;
+        $helper->toolbar_btn = $this->initToolbar();
+
+        return $helper;
+    }
+
+    public function validateForm($data)
+    {
+        return true;
```
Comment (mickaelandrieu) [inline controllers/admin/AdminLinkWidgetController.php]: hmm...

File: src/CmsBlock.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        if (is_array($this->content)) {
+            $this->content = json_encode($this->content);
+        }
+
+        if (!$this->position) {
+            $this->position = 1;
+        }
+
+        $return = parent::add($auto_date, $null_values);
```
Comment (mickaelandrieu) [inline src/CmsBlock.php]: $return is supposed to be boolean or integer ?  I'd like to have an explicit variable name here like $additionOk or $additionStatusCode.

File: src/CmsBlockPresenter.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        $this->link = $link;
+        $this->language = $language;
+    }
+
+    public function present(CmsBlock $cmsBlock)
+    {
+        return [
+            'id' => $cmsBlock->id,
+            'title' => $cmsBlock->name[$this->language->id],
+            'hook' => (new Hook($cmsBlock->id_hook))->name,
```
Comment (mickaelandrieu) [inline src/CmsBlockPresenter.php]: `hook_name` is more explicit, what do you think ?

File: src/CmsBlockPresenter.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            'position' => $cmsBlock->position,
+            'links' => $this->makeLinks($cmsBlock->content)
+        ];
+    }
+
+    private function makeLinks($content)
```
Comment (mickaelandrieu) [inline src/CmsBlockPresenter.php]: Why not use makeCmsLinks directly ? Do you plan to complete this function later ?

File: src/CmsBlockRepository.php [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+                    'id_hook' => $block['id_hook'],
+                    'hook_name' => $block['hook_name'],
+                    'hook_title' => $block['hook_title'],
+                    'hook_description' => $block['hook_description'],
+                    'blocks' => [],
+                ];
+            }
+        }
+
+        foreach ($blocks as $block) {
```
Comment (mickaelandrieu) [inline src/CmsBlockRepository.php]: You may probably do this treatment in the previous ``foreach``, also multiple unset have better perfs.

For instance

```php
unset($block['id_hook'], $block['hook_name], ...);
```

make only one function call.

**Comment by mickaelandrieu** [comment]
By the way, good job dude :smile_cat:

File: src/CmsBlockPresenter.php (hunk 7 above)
Comment (julienbourdeau) [inline src/CmsBlockPresenter.php]: Yeah it's meant to handle more links: search results, categories, module pages,...

File: src/CmsBlockPresenter.php (hunk 6 above)
Comment (julienbourdeau) [inline src/CmsBlockPresenter.php]: :+1:

**Pull request merged**